chart.js.d.ts

```diff
@@ -0,0 +1,605 @@
+    afterSetDimensions?:          (scale: Scale) => void; // Check required
+    beforeDataLimits?:            (scale: Scale) => void; // Check required
+    afterDataLimits?:             (scale: Scale) => void; // Check required
+    beforeBuildTicks?:            (scale: Scale) => void; // Check required
+    afterBuildTicks?:             (scale: Scale) => void; // Check required
+    beforeTickToLabelConversion?: (scale: Scale) => void; // Check required
+    afterTickToLabelConversion?:  (scale: Scale) => void; // Check required
+    beforeCalculateTickRotation?: (scale: Scale) => void; // Check required
+    afterCalculateTickRotation?:  (scale: Scale) => void; // Check required
+    beforeFit?:                   (scale: Scale) => void; // Check required
+    afterFit?:                    (scale: Scale) => void; // Check required
+    afterUpdate?:                 (scale: Scale) => void; // Check required
+    // @formatter:on
+    gridLines?: Array<ScaleGridLinesOptions>;
+    scaleLabel?: Array<ScaleLabelOptions>;
+    ticks?: Array<ScaleTickOptions>;
+  }
+
+  interface ScaleGridLinesOptions {
+    display?: boolean
+    color?: string
+    lineWidth?: number
+    drawOnChartArea?: boolean
+    drawTicks?: boolean
+    zeroLineWidth?: number
+    zeroLineColor?: string
+    offsetGridLines?: boolean
+  }
+
+  interface ScaleLabelOptions {
+    display?: boolean
+    labelString?: string
+    fontColor?: string
+    fontFamily?: string
+    fontSize?: number
+    fontStyle?: string
+  }
+
+  interface ScaleTickOptions {
+    beginAtZero?: boolean
+    fontColor?: string
+    fontFamily?: string
+    fontSize?: number
+    fontStyle?: string
+    maxRotation?: number
+    minRotation?: number
+    maxTicksLimit?: number
+    padding?: number
+    mirror?: boolean
+    reverse?: boolean
+    display?: boolean
+    suggestedMin?: number
+    suggestedMax?: number
+    min?: number
+    max?: number
+    autoSkip?: boolean
+    callback?: (value: number) => string // Check required
+  }
+
+  interface LinearScaleOptions extends ScaleOptions {
+    ticks?: Array<ScaleTickOptions & {
+      stepSize?: number
+    }>
+  }
+
+  // Check Required, do the properties (starting at left) belong in Scale class and top level?
+  abstract class Scale {
+
+    abstract determineDataLimits(): void;
+
+    abstract buildTicks(): void
+
+    abstract getLabelForIndex(index: number, dataSetIndex: number): number
+
+    abstract getPixelForTick(index: number, includeOffset: boolean): number
+
+    abstract getPixelForValue(value: number, index: number, datasetIndex: number, includeOffset: boolean): number
+
+    convertTicksToLabels(): void
+
+    calculateTickRotation(): void
+
+    fit(): void
+
+    draw(chartArea: ChartArea): void // Is ChartArea independent object or part of ChartInstance?
+
+    isHorizontal(): boolean
+
+    getRightValue(dataValue: Object): any // NaN is a possibility here (is a more constricting type possible?)
```

Comment:
I think anything could be returned because this could be overridden in other scales which may have custom data types